File: src/org/elixir_lang/Module.java
package org.elixir_lang;

import com.intellij.openapi.util.Condition;
import com.intellij.openapi.util.text.StringUtil;
import org.apache.commons.lang.ArrayUtils;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;

import java.util.Arrays;
import java.util.Collection;

public class Module {
    public static class IsNestedUnder implements Condition<String> {
        /*
         * Fields
         */

        @NotNull
        private java.util.List<String> splitModuleName;

        /*
         * Constructors
         */

        public IsNestedUnder(@NotNull String moduleName) {
            this.splitModuleName = split(moduleName);
        }

        /*
         * Public Instance Methods
         */

        @Override
        public boolean value(@NotNull String maybeStartsWithModuleName) {
            java.util.List<String> splitMaybStartsIwthModuleName = split(maybeStartsWithModuleName);
            boolean isNestedUnder = true;

            if (splitMaybStartsIwthModuleName.size() > splitModuleName.size()) {
                for (int i = 0; i < splitModuleName.size(); i++) {
                    if (!splitMaybStartsIwthModuleName.get(i).equals(splitModuleName.get(i))) {
                        isNestedUnder = false;

                        break;
                    }
                }
            } else {
                isNestedUnder = false;
            }

            return isNestedUnder;
        }
    }

    /*
     * Constants
     */

    private static final String SEPARATOR = ".";

    /*
     * Public Static Methods
     */

    /**
     * Emulates Module.concat/1
     * @return
     */
    @Contract(pure = true)
    @NotNull
    public static String concat(@NotNull String... aliases) {
        return StringUtil.join(aliases, SEPARATOR);
    }

    @Contract(pure = true)
    @NotNull
    public static String concat(@NotNull Collection<String> aliases) {
        return StringUtil.join(aliases, SEPARATOR);
    }

    /**
     * Emulates Module.split/1
     */
    @Contract(pure = true)
    @NotNull
    public static java.util.List<String> split(@NotNull String name) {
        return StringUtil.split(name, SEPARATOR);
    }
}

File: src/org/elixir_lang/Module.kt
package org.elixir_lang

import com.intellij.openapi.util.Condition
import org.jetbrains.annotations.Contract
import kotlin.collections.List

object Module {
    private const val SEPARATOR = "."

    class IsNestedUnder (moduleName: String) : Condition<String> {
        private val splitModuleName: List<String> = split(moduleName)

        override fun value(maybeStartsWithModuleName: String): Boolean {
            val splitMaybeStartsWithModuleName = split(maybeStartsWithModuleName)
            var isNestedUnder = true

            if (splitMaybeStartsWithModuleName.size > splitModuleName.size) {
                for (i in splitModuleName.indices) {
                    if (splitMaybeStartsWithModuleName[i] != splitModuleName[i]) {
                        isNestedUnder = false

                        break
                    }
                }
            } else {
                isNestedUnder = false
            }

            return isNestedUnder
        }
    }

    /**
     * Emulates Module.concat/1
     */
    @Contract(pure = true)
    @JvmStatic
    fun concat(aliases: Collection<String>): String = aliases.joinToString(SEPARATOR)

    /**
     * Emulates Module.split/1
     */
    @Contract(pure = true)
    @JvmStatic
    fun split(name: String): List<String> = name.split(SEPARATOR)

    fun prefix(name: String): List<String> = split(name).dropLast(1)

    fun isRelative(ancestors: List<String>, descendant: String): Boolean = relative(ancestors, descendant).isNotEmpty()

    fun relative(ancestors: List<String>, descendant: String): List<String> {
        val descendants = split(descendant)
        return relative(ancestors, descendants)
    }

    fun relative(ancestors: List<String>, descendants: List<String>): List<String> {
        return if (ancestors.size < descendants.size &&
                ancestors.zip(descendants).all { (ancestor, descendent) -> ancestor == descendent }) {
            descendants.subList(ancestors.size, descendants.size)
        } else {
            emptyList()
        }
    }
}
